### PR TITLE
fix: timeout on Non-AVR boards; feat: Use yield() in busy wait loops

### PR DESCRIPTION
**Pull request by greezybacon**
Also, be more scientific about the timeouts rather than using a set number of loops, the time of which depends on the value of the `MRFC522_SPICLOCK` as well as `F_CPU`. Instead, use a set number of milliseconds as a deadline.

In my project, I'm using the EventResponder to perform activities during busy wait loops like some of the loops in the library. Using `yield()` allows the timing on the activities to be more consistent. Plus, I'm also putting the MCU into a lower power mode in as part of the EventResponder/idle loop which allows for better power and heat management in `delay` and wait loops.

On the Teensy platform, a call to `millis()` is free as it's a register lookup. It's also a register lookup on the Arduino platform with a very small amount of overhead. All in all, for folks not using `yield()` I would think this edit wouldn't provide a negative impact.

edit by @Rotzbua : The previous timeout code was designed for AVR based Arduinos. Now it should work properly on Non-AVR boards. So I think it is also a bug fix.

| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | yes
| Doc update?   | no
| BC breaks?    | no 
| Deprecations? | no
| Fixed tickets | not sure

### CODE REVIEW

File: src/MFRC522.cpp [hunk 1]
```diff
@@ -167,7 +167,9 @@ MFRC522::StatusCode MFRC522::PCD_CalculateCRC(	byte *data,		///< In: Pointer to
 	// TODO check/modify for other architectures than Arduino Uno 16bit
```
Comment (Rotzbua) [inline src/MFRC522.cpp:167]: update doc please

File: src/MFRC522.cpp [hunk 2]
```diff
@@ -481,18 +487,22 @@ MFRC522::StatusCode MFRC522::PCD_CommunicateWithPICC(	byte command,		///< The co
 	// In PCD_Init() we set the TAuto flag in TModeReg. This means the timer automatically starts when the PCD stops transmitting.
 	// Each iteration of the do-while-loop takes 17.86μs.
```
Comment (Rotzbua) [inline src/MFRC522.cpp:488]: update doc please

File: src/MFRC522.cpp [hunk 3]
```diff
@@ -177,7 +179,10 @@ MFRC522::StatusCode MFRC522::PCD_CalculateCRC(	byte *data,		///< In: Pointer to
 			result[1] = PCD_ReadRegister(CRCResultRegH);
 			return STATUS_OK;
 		}
+		yield();
 	}
+	while (millis() < deadline);
```
Comment (Rotzbua) [inline src/MFRC522.cpp:184]: please use typecast for `millis` to `uint32_t`

File: src/MFRC522.cpp [hunk 4]
```diff
@@ -481,18 +487,22 @@ MFRC522::StatusCode MFRC522::PCD_CommunicateWithPICC(	byte command,		///< The co
...
 		byte n = PCD_ReadRegister(ComIrqReg);	// ComIrqReg[7..0] bits are: Set1 TxIRq RxIRq IdleIRq HiAlertIRq LoAlertIRq ErrIRq TimerIRq
 		if (n & waitIRq) {					// One of the interrupts that signal success has been set.
 			break;
 		}
 		if (n & 0x01) {						// Timer interrupt - nothing received in 25ms
 			return STATUS_TIMEOUT;
 		}
+		yield();
 	}
+	while (millis() < deadline);
```
Comment (Rotzbua) [inline src/MFRC522.cpp:502]: please use typecast for `millis` to `uint32_t`

**Review by Rotzbua — CHANGES_REQUESTED** [review]

File: src/MFRC522.cpp (hunk 3 above)
Comment (greezybacon) [reply src/MFRC522.cpp:184]: `millis` is `unsigned long` on the Arduino platform. Would you prefer `unsigned long` over `uint32_t`?

File: src/MFRC522.cpp (hunk 1 above)
Comment (greezybacon) [reply src/MFRC522.cpp:167]: Would you prefer to have `PCD_CalculateCRC` receive a `timeout` parameter with a default of `89` for backwards compatibility?

File: src/MFRC522.cpp (hunk 2 above)
Comment (greezybacon) [reply src/MFRC522.cpp:488]: Would you prefer to have `PCD_CommunicateWithPICC` receive a `timeout` parameter with a default of 36 for backwards compatibility?

File: src/MFRC522.cpp (hunk 3 above)
Comment (Rotzbua) [reply src/MFRC522.cpp:184]: i prefer the `uint32_t` over the generic `unsigned long`

File: src/MFRC522.cpp (hunk 1 above)
Comment (Rotzbua) [reply src/MFRC522.cpp:167]: i think there is no real use case for a new parameter?
just remove the `todo` because you solved it 👍

Comment (Rotzbua) [reply src/MFRC522.cpp:167]: and remove the loop comment to a more suitable comment thanks

**Comment by greezybacon** [comment]
@Rotzbua I made the changes you requested. Let me know what you think of them and if you'd like other changes.

File: src/MFRC522.cpp [hunk 5]
```diff
@@ -177,7 +179,10 @@ MFRC522::StatusCode MFRC522::PCD_CalculateCRC(	byte *data,		///< In: Pointer to
 			result[1] = PCD_ReadRegister(CRCResultRegH);
 			return STATUS_OK;
 		}
+		yield();
 	}
+	while (static_cast<uint32_t> (millis()) < deadline);
+
 	// 89ms passed and nothing happend. Communication with the MFRC522 might be down.
```
Comment (Rotzbua) [inline src/MFRC522.cpp:186]: `happened`

File: src/MFRC522.cpp [hunk 6]
```diff
@@ -477,22 +483,32 @@ MFRC522::StatusCode MFRC522::PCD_CommunicateWithPICC(	byte command,		///< The co
...
 		if (n & 0x01) {						// Timer interrupt - nothing received in 25ms
 			return STATUS_TIMEOUT;
 		}
+		yield();
 	}
-	// 35.7ms and nothing happend. Communication with the MFRC522 might be down.
-	if (i == 0) {
+	while (static_cast<uint32_t> (millis()) < deadline);
+
+	// 36ms and nothing happend. Communication with the MFRC522 might be down.
```
Comment (Rotzbua) [inline src/MFRC522.cpp:510]: `happened`

File: src/MFRC522.cpp [hunk 7]
```diff
@@ -477,22 +483,32 @@ MFRC522::StatusCode MFRC522::PCD_CommunicateWithPICC(	byte command,		///< The co
...
-	for (i = 2000; i > 0; i--) {
+	// In PCD_Init() we set the TAuto flag in TModeReg. This means the timer
+	// automatically starts when the PCD stops transmitting.
+	//
+	// Wait here for the command to complete. The bits specified in the
+	// `waitIRq` parameter define what bits constitute a completed command.
+	// When they are set in the ComIrqReg register, then the command is
+	// considered complete. If the command is not indicated as complete in
+	// ~36ms, then consider the command as timed out.
+	uint32_t deadline = millis() + 36;
```
Comment (Rotzbua) [inline src/MFRC522.cpp:494]: could be `const uint32_t`

File: src/MFRC522.cpp [hunk 8]
```diff
@@ -163,11 +163,13 @@ MFRC522::StatusCode MFRC522::PCD_CalculateCRC(	byte *data,		///< In: Pointer to
 	PCD_WriteRegister(FIFODataReg, length, data);	// Write data to the FIFO
 	PCD_WriteRegister(CommandReg, PCD_CalcCRC);		// Start the calculation
 	
-	// Wait for the CRC calculation to complete. Each iteration of the while-loop takes 17.73μs.
-	// TODO check/modify for other architectures than Arduino Uno 16bit
+	// Wait for the CRC calculation to complete. Check for the register to
+	// indicate that the CRC calculation is complete in a loop. If the
+	// calculation is not indicated as complete in ~90ms, then time out
+	// the operation.
+	uint32_t deadline = millis() + 89;
```
Comment (Rotzbua) [inline src/MFRC522.cpp:170]: could be `const uint32_t`

**Review by Rotzbua — COMMENTED** [review]
nice to have changes

**Comment by greezybacon** [comment]
@Rotzbua done.

**Pull request merged**